File: data_structures/python/linked-list.py
import sys


class ListNode:
    def __init__(self, data: int) -> None:
        self.data = data
        self.next: ListNode = None


class LinkedList:
    def __init__(self) -> None:
        self.first: ListNode = None

    def print_list(self):
        if self.first is None:
            print("List is empty")
            return

        curr = self.first
        while curr:
            sys.stdout.write(str(curr.data) + " ")
            curr = curr.next
        sys.stdout.write("\n")

    def push_front(self, x: int):
        newNode = ListNode(x)
        newNode.next = self.first
        self.first = newNode

    def push_back(self, x: int):
        newNode = ListNode(x)
        if not self.first:
            self.first = newNode
            return

        curr = self.first
        while curr.next:
            curr = curr.next

        curr.next = newNode

    def delete(self, x: int):
        curr = self.first
        prev = None

        while curr and curr.data != x:
            prev = curr
            curr = curr.next

        if not curr:
            print(f"There is no {x} in list.")

        elif curr == self.first:
            self.first = curr.next
            del curr

        else:
            prev.next = curr.next
            del curr

    def clear(self):
        while self.first:
            curr = self.first
            self.first = self.first.next
            del curr

    def reverse(self):
        if self.first is None or self.first.next is None:
            return

        prev = None
        curr = self.first
        prec = self.first.next

        while prec:
            curr.next = prev
            prev = curr
            curr = prec
            prec = prec.next

        curr.next = prev
        self.first = curr


if __name__ == "__main__":
    list = LinkedList()
    # 建立LinkedList的object
    list.print_list()
    # 目前list是空的
    list.delete(4)
    # list是空的, 沒有4
    list.push_back(5)
    # list: 5
    list.push_back(3)
    # list: 5 3
    list.push_front(9)
    # list: 9 5 3
    list.print_list()
    # 印出:  9 5 3
    list.push_back(4)
    # list: 9 5 3 4
    list.delete(9)
    # list: 5 3 4
    list.print_list()
    # 印出:  5 3 4
    list.push_front(8)
    # list: 8 5 3 4
    list.print_list()
    # 印出:  8 5 3 4
    list.reverse()
    # list: 4 3 5 8
    list.print_list()
    # 印出:  4 3 5 8
    list.clear()
    # 清空list
    list.print_list()
    # 印出: List is empty.

File: data_structures/c_plus_plus/linked_list.cpp
#include <iostream>

class LinkedList;

class ListNode
{
private:
    int data;
    ListNode *next;

public:
    ListNode() : data(0), next(0){};
    ListNode(int a) : data(a), next(0){};
    friend class LinkedList;
};

class LinkedList
{
private:
    ListNode *first;

public:
    LinkedList() : first(0){};
    void PrintList();
    void Push_front(int x);
    void Push_back(int x);
    void Delete(int x);
    void Clear();
    void Reverse();
};

void LinkedList::PrintList()
{
    if (first == 0)
    {
        std::cout << "List is empty" << std::endl;
        return;
    }

    ListNode *curr = first;
    while (curr != 0)
    {
        std::cout << curr->data << " ";
        curr = curr->next;
    }
    std::cout << std::endl;
}

void LinkedList::Push_front(int x)
{
    ListNode *newNode = new ListNode(x);
    newNode->next = first;
    first = newNode;
}

void LinkedList::Push_back(int x)
{
    ListNode *newNode = new ListNode(x);

    if (first == 0)
    {
        first = newNode;
        return;
    }

    ListNode *current = first;
    while (current->next != 0)
    {
        current = current->next;
    }
    current->next = newNode;
}

void LinkedList::Delete(int x)
{
    ListNode *curr = first,
             *prev = 0;

    while (curr != 0 && curr->data != x)
    {
        prev = curr;
        curr = curr->next;
    }

    if (curr == 0)
    {
        std::cout << "There is no " << x << " in list." << std::endl;
    }
    else if (curr == first)
    {
        first = curr->next;
        delete curr;
        curr = 0;
        return;
    }
    else
    {
        prev->next = curr->next;
        delete curr;
        curr = 0;
        return;
    }
}

void LinkedList::Clear()
{
    while (first != 0)
    {
        ListNode *curr = first;
        first = first->next;
        delete curr;
        curr = 0;
    }
}

void LinkedList::Reverse()
{
    if (first == 0 || first->next == 0)
    {
        return;
    }

    ListNode *prev = 0,
             *curr = first,
             *prec = first->next;

    while (prec != 0)
    {
        curr->next = prev;
        prev = curr;
        curr = prec;
        prec = prec->next;
    }

    curr->next = prev;
    first = curr;
}

int main()
{

    LinkedList list;    // 建立LinkedList的object
    list.PrintList();   // 目前list是空的
    list.Delete(4);     // list是空的, 沒有4
    list.Push_back(5);  // list: 5
    list.Push_back(3);  // list: 5 3
    list.Push_front(9); // list: 9 5 3
    list.PrintList();   // 印出:  9 5 3
    list.Push_back(4);  // list: 9 5 3 4
    list.Delete(9);     // list: 5 3 4
    list.PrintList();   // 印出:  5 3 4
    list.Push_front(8); // list: 8 5 3 4
    list.PrintList();   // 印出:  8 5 3 4
    list.Reverse();     // list: 4 3 5 8
    list.PrintList();   // 印出:  4 3 5 8
    list.Clear();       // 清空list
    list.PrintList();   // 印出: List is empty.

    return 0;
}